### PR TITLE
support cron for ehpa

### PR DESCRIPTION
**Pull request by kitianFresh**

### CODE REVIEW

File: autoscaling/v1alpha1/ehpa_types.go [hunk 1]
```diff
@@ -66,6 +71,23 @@ type EffectiveHorizontalPodAutoscalerSpec struct {
...
+	Name 			string `json:"name,omitempty"`
+	// Description is the description of the cron
+	Description     string 	`json:"description,omitempty"`
+	// TimeZone is the time zone of this cron schedule running in
+	TimeZone		*string `json:"timezone,omitempty"`
+	// Start is a crontab format, define the cron schedule start, when the cron start is triggered, hpa will reconcile targetRef to scale to the TargetReplicas continuously.
+	Start  		string	`json:"start,omitempty"`
+	// End is a crontab format, define the cron schedule end, when it ended, EHPA will stop to trigger hpa to scale.
+	// Between the start and end, EHPA will keep replicas of the targetRef to equal to TargetReplicas, a scaling is a process rather than a instant action, make it as a final consistent system.
+	End 		string	`json:"end,omitempty"`
```
Comment (qmhu) [inline autoscaling/v1alpha1/ehpa_types.go:86]: How to handle ?
start != nil and end == nil 
start == nil and end != nil

Comment (kitianFresh) [reply autoscaling/v1alpha1/ehpa_types.go:86]: > How to handle ? start != nil and end == nil start == nil and end != nil

WebHook validate the start and end, start & end must not be empty.

File: autoscaling/v1alpha1/ehpa_types.go [hunk 2]
```diff
@@ -66,6 +71,23 @@ type EffectiveHorizontalPodAutoscalerSpec struct {
 	Prediction *Prediction `json:"prediction,omitempty"`
 }
 
+// CronScale defines the cron scale info
+type CronScale struct {
```
[the remark this answers is not in the document]
Comment (kitianFresh) [reply autoscaling/v1alpha1/ehpa_types.go:75]: > `CronScale` or `CronSpec`?

Ok, use CronSpec to keep the same style with MetricSpec

File: autoscaling/v1alpha1/ehpa_types.go [hunk 3]
```diff
@@ -66,6 +71,23 @@ type EffectiveHorizontalPodAutoscalerSpec struct {
...
+	// Name is the identifier of this cron scale. name must be unique in the same ehpa
+	Name 			string `json:"name,omitempty"`
+	// Description is the description of the cron
+	Description     string 	`json:"description,omitempty"`
```
[the remark this answers is not in the document]
Comment (kitianFresh) [reply autoscaling/v1alpha1/ehpa_types.go:79]: ok

File: autoscaling/v1alpha1/ehpa_types.go [hunk 4]
```diff
@@ -66,6 +71,23 @@ type EffectiveHorizontalPodAutoscalerSpec struct {
...
+	// TimeZone is the time zone of this cron schedule running in
+	TimeZone		*string `json:"timezone,omitempty"`
```
Comment (mfanjie) [inline autoscaling/v1alpha1/ehpa_types.go:81]: one question: why TimeZone is *string while other fields are string.

Comment (kitianFresh) [reply autoscaling/v1alpha1/ehpa_types.go:84]: > one question: why TimeZone is *string while other fields are string.

it has default value when user do not specify it. When it is null, system use utc time zone

**Pull request merged**